plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.Nullable;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.plugin.insights.core.exporter.QueryInsightsExporter;
+import org.opensearch.plugin.insights.core.exporter.QueryInsightsExporterType;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.plugin.insights.settings.QueryInsightsSettings;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting data related to
+ * search queries, based on certain dimensions.
+ *
+ * @param <R> The type of record that stores in the service
+ * @param <S> The type of Collection that holds the aggregated data
+ * @param <E> The type of exporter that exports the aggregated and processed data
+ *
+ * @opensearch.internal
+ */
+public abstract class QueryInsightsService<R extends SearchQueryRecord<?>, S extends Collection<R>, E extends QueryInsightsExporter<R>>
```

Comment:
Do we really need S in the type definition? Can we directly define this as Collection<R>

Reply:
The subclasses need this to know the "type of collection" so that the they can do related operations based on the type. For example, in the top N queries case, the sub class needs this Collection to be a priority queue, this needs to be provided as S.

Reply:
Actually using S directly has advantage as the return type will not be Collection<R> and caller can take advantage of more methods in S instead of limiting to the super type Collection<R>.

Although, if subclass is not planning to do more than Collection<R> based operations using this class, probably it can be removed